src/infra/dex/balancer/dto.rs

```diff
@@ -1,58 +1,208 @@
 use {
     crate::{
-        domain::{auction, dex, order},
+        domain::{dex, eth, order},
+        infra::dex::balancer::Error,
         util::serialize,
     },
+    bigdecimal::BigDecimal,
     ethereum_types::{H160, H256, U256},
+    number::conversions::{big_decimal_to_u256, u256_to_big_decimal},
     serde::{Deserialize, Serialize},
     serde_with::serde_as,
 };
 
-#[derive(Debug, Serialize)]
+pub mod get_swap_paths_query {
+    /// Get swap quote from the SOR v2 for the V2 vault.
+    pub const QUERY: &str = r#"
+query sorGetSwapPaths($callDataInput: GqlSwapCallDataInput!, $chain: GqlChain!, $queryBatchSwap: Boolean!, $swapAmount: AmountHumanReadable!, $swapType: GqlSorSwapType!, $tokenIn: String!, $tokenOut: String!, $useVaultVersion: Int) {
+    sorGetSwapPaths(
+        callDataInput: $callDataInput,
+        chain: $chain,
+        queryBatchSwap: $queryBatchSwap,
+        swapAmount: $swapAmount,
+        swapType: $swapType,
+        tokenIn: $tokenIn,
+        tokenOut: $tokenOut,
+        useVaultVersion: $useVaultVersion
+    ) {
+        tokenAddresses
+        swaps {
+            poolId
+            assetInIndex
+            assetOutIndex
+            amount
+            userData
+        }
+        swapAmountRaw
+        returnAmountRaw
+        tokenIn
+        tokenOut
+    }
+}
+    "#;
+}
+
+#[derive(Serialize)]
 #[serde(rename_all = "camelCase")]
-pub enum OrderKind {
-    Sell,
-    Buy,
+pub struct Query<'a> {
+    query: &'a str,
+    variables: Variables,
+}
+
+impl Query<'_> {
+    pub fn from_domain(
+        order: &dex::Order,
+        slippage: &dex::Slippage,
+        chain_id: eth::ChainId,
+        contract_address: eth::ContractAddress,
+        query_batch_swap: bool,
+    ) -> Result<Self, Error> {
+        let variables = Variables {
+            call_data_input: CallDataInput {
+                deadline: None,
+                receiver: contract_address.0,
+                sender: contract_address.0,
+                slippage_percentage: slippage.as_factor().clone(),
+            },
+            chain: Chain::from_domain(chain_id)?,
+            query_batch_swap,
+            swap_amount: EtherAmount::from_wei(&order.amount.get()),
+            swap_type: SwapType::from_domain(order.side),
+            token_in: order.sell.0,
+            token_out: order.buy.0,
+            use_vault_version: VaultVersion::V2 as u8,
+        };
+        Ok(Self {
+            query: get_swap_paths_query::QUERY,
+            variables,
+        })
+    }
+}
+
+/// Ether amount refers to the SOR API V3's `AmountHumanReadable` type.
+#[serde_as]
+#[derive(Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct EtherAmount(BigDecimal);
+
+impl EtherAmount {
+    pub fn from_wei(wei: &U256) -> EtherAmount {
+        Self(u256_to_big_decimal(wei) / BigDecimal::from(10_u64.pow(18)))
+    }
+
+    pub fn value(&self) -> BigDecimal {
+        self.0.clone()
```

Comment:
Could we pass a reference to it? we can leave the clone to the caller (if the caller needs the clone or not is up the caller)